### PR TITLE
Change the request and the response of `Posts`

### LINKED ISSUE #498: Change how post are responded
## Feature description 

In current version, the post response returns whole content includes all the `PostReactions` and `UserAnswers`. In my opinion, both of them in `PostResponse` could be replaced with the total count. As a result, the post response will be much smaller and much more fast when querying. Then, `PostReactions` can also be paginated when querying whole reactions of the post.

## Implementation proposal

Change the post response struct, say:
```proto
message QueryPostResponse {
  desmos.posts.v1beta1.Post post = 1;

  // Answers field show the total count of each answers from user
  map<string, int> answers = 2;

  // Reactions field shows the total count of each reactions from user 
  map<string, int> reactions = 3;

  repeated string children = 4;
}
```
Thoughts on it? @RiccardoM @bragaz

### PR DESCRIPTION
**Pull request by dadamu**
## Description

This PR is the implementation of #498 and #500. 
The pagination feature for `PostReaction` and  `PostChildren` will be implemented by other issues.
Closes #498 #500

## Checklist
- [x] Targeted PR against correct branch.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [x] Wrote unit tests.
- [ ] Wrote integration tests (simulation & CLI). 
- [x] Updated the documentation. 
- [x] Added an entry to the `CHANGELOG.md` file.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: x/staging/posts/client/cli/query.go [hunk 1]
```diff
@@ -75,12 +72,12 @@ func GetCmdQueryPost() *cobra.Command {
 func GetCmdQueryPosts() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "posts",
-		Short: "Query posts with optional filters",
+		Short: "Query posts with the optional subspace",
```
Comment (bragaz) [inline x/staging/posts/client/cli/query.go:75]:
```suggestion
		Short: "Query posts with optional subspace",
```

File: x/staging/posts/client/cli/query.go [hunk 2]
```diff
@@ -75,12 +72,12 @@ func GetCmdQueryPost() *cobra.Command {
...
 		Long: strings.TrimSpace(
-			fmt.Sprintf(`Query for paginated posts that match optional filters:
+			fmt.Sprintf(`Query for paginated posts that match the optional subspace:
```
Comment (bragaz) [inline x/staging/posts/client/cli/query.go:77]:
```suggestion
			fmt.Sprintf(`Query for paginated posts that match an optional subspace:
```

**Review by bragaz — APPROVED** [review]
The whole implementation looks good for me.
Let's wait for @RiccardoM to double check.

**Comment by dadamu** [comment]
@bragaz 
I found that the iterating function getting all the posts sorted by the index. Just remove it and all the index related thingy.

File: CHANGELOG.md [hunk 3]
```diff
@@ -11,6 +11,7 @@
 - Added the ability to paginate the user poll answers ([#478](https://github.com/desmos-labs/desmos/issues/478))
 - Added the logging of some interesting operations ([#483](https://github.com/desmos-labs/desmos/issues/483))
 - Make `x/profiles` DTag case-insensitive ([#492](https://github.com/desmos-labs/desmos/issues/492))
+- Changed how posts are queried ([#499](https://github.com/desmos-labs/desmos/issues/499))
```
Comment (RiccardoM) [inline CHANGELOG.md:14]:
```suggestion
- Improved the posts query response ([#499](https://github.com/desmos-labs/desmos/issues/499))
```

File: proto/desmos/posts/v1beta1/query.proto [hunk 4]
```diff
@@ -51,49 +51,18 @@ service Query {
...
-    (gogoproto.jsontag) = "parent_id",
-    (gogoproto.moretags) = "yaml:\"id\""
-  ];
-  ;
-  google.protobuf.Timestamp creation_time = 4 [ (gogoproto.stdtime) = true ];
-  string subspace = 5;
-  string creator = 6;
-  repeated string hashtags = 7;
+  
+  string subspace = 1;
```
Comment (RiccardoM) [inline proto/desmos/posts/v1beta1/query.proto:57]:
```suggestion
  // Subspace must contain the ID of the subspace to which query the posts for. 
  // Providing an empty value will return an error.
  string subspace = 1;
```

File: x/staging/posts/client/cli/query.go [hunk 5]
```diff
@@ -75,12 +72,12 @@ func GetCmdQueryPost() *cobra.Command {
 func GetCmdQueryPosts() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "posts",
-		Short: "Query posts with optional filters",
+		Short: "Query posts with optional subspace",
```
Comment (RiccardoM) [inline x/staging/posts/client/cli/query.go:75]:
```suggestion
		Short: "Query posts with optional pagination",
```

File: x/staging/posts/client/cli/query.go [hunk 6]
```diff
@@ -75,12 +72,12 @@ func GetCmdQueryPost() *cobra.Command {
 func GetCmdQueryPosts() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "posts",
```
Comment (RiccardoM) [inline x/staging/posts/client/cli/query.go:74]: I think it's much better to require the user to specify a subspace. This will block them from being able to performing high-intensive queries like fetching all the posts. 

```suggestion
		Use:   "posts [subspace]",
		Args: os.ExactArgs(1),
```

File: x/staging/posts/keeper/grpc_query.go [hunk 7]
```diff
@@ -16,98 +15,38 @@ import (
...
-				filteredPosts = append(filteredPosts, k.getPostResponse(ctx, post))
-			}
-
-			return true, nil
+		if accumulate {
+			posts = append(posts, post)
 		}
 
-		return false, nil
+		return true, nil
```
Comment (RiccardoM) [inline x/staging/posts/keeper/grpc_query.go:39]: Why was this changed? Shouldn't it be kept to `false` since it represents the `stop`?

File: x/staging/posts/keeper/grpc_query.go [hunk 8]
```diff
@@ -16,98 +15,38 @@ import (
...
-	return types.QueryPostResponse{
-		Post:        post,
-		UserAnswers: answers,
-		Reactions:   postReactions,
-		Children:    childrenIDs,
-	}
-}
-
+// Posts implements the Query/Posts gRPC method
 func (k Keeper) Posts(goCtx context.Context, req *types.QueryPostsRequest) (*types.QueryPostsResponse, error) {
```
Comment (RiccardoM) [inline x/staging/posts/keeper/grpc_query.go:19]: I think that here we should require the user to give us a valid subspace identifier, blocking them from querying the whole set of posts. There should not be any reason to do that anyway.

File: x/staging/posts/keeper/keeper.go [hunk 9]
```diff
@@ -65,19 +61,9 @@ func (k Keeper) SavePost(ctx sdk.Context, post types.Post) {
...
-
-		postIndex = types.PostIndex{Value: postIndex.Value + 1}
-
-		// Save the new incremental ID of the post and update the total number of posts
-		store.Set(types.PostIndexedIDStoreKey(post.PostID), k.cdc.MustMarshalBinaryBare(&postIndex))
-		store.Set(types.PostTotalNumberPrefix, k.cdc.MustMarshalBinaryBare(&postIndex))
+	// Save the query key if the key does not exist
+	if !store.Has(types.PostQueryStoreKey(post.Subspace, post.PostID)) {
+		store.Set(types.PostQueryStoreKey(post.Subspace, post.PostID), []byte(post.PostID))
 	}
```
Comment (RiccardoM) [inline x/staging/posts/keeper/keeper.go:67]: This can and should be simplified to avoid calling a method twice: 
```suggestion
	subspaceKey := types.PostQueryStoreKey(post.Subspace, post.PostID)
	if !store.Has(subspaceKey) {
		store.Set(subspaceKey, []byte(post.PostID))
	}
```

File: x/staging/posts/simulation/utils.go [hunk 10]
```diff
@@ -89,7 +87,7 @@ func RandomPostData(r *rand.Rand, accs []simtypes.Account) PostData {
 	post := types.NewPost(
 		"",
 		RandomPostID(r),
-		RandomMessage(r)+RandomHashtag(r),
```
Comment (RiccardoM) [inline x/staging/posts/simulation/utils.go:92]: Why has this been deleted?

File: x/staging/posts/types/keys.go [hunk 11]
```diff
@@ -69,16 +61,21 @@ func PostStoreKey(id string) []byte {
...
 
+// PostQuerySubspacePrefix returns the prefix used to store all the post id inside the subspace having the given id
+func PostQuerySubspacePrefix(subspace string) []byte {
+	return append(PostQueryStorePrefix, []byte(subspace)...)
+}
+
+// PostQueryStoreKey returns the store key to store the post id containing the given data
+func PostQueryStoreKey(subspace string, id string) []byte {
+	return append(PostQuerySubspacePrefix(subspace), []byte(id)...)
+}
```
Comment (RiccardoM) [inline x/staging/posts/types/keys.go:77]: I think we can rename those to be more clear

```suggestion
// SubspacePostPrefix returns the prefix used to store all the posts present inside the subspace having the given id
func SubspacePostsPrefix(subspace string) []byte {
	return append(SubspacePostPrefix, []byte(subspace)...)
}

// SubspacePostKey returns the key used to associate the post with the given id to the subspace with the provided id
func SubspacePostKey(subspace string, id string) []byte {
	return append(PostQuerySubspacePrefix(subspace), []byte(id)...)
}
```

File: x/staging/posts/types/keys.go [hunk 12]
```diff
@@ -29,25 +29,17 @@ const (
...
 	shortCodeRegEx = regexp.MustCompile(`:[a-z0-9+-]([a-z0-9\d_-])*:`)
-	hashtagRegEx   = regexp.MustCompile(`[^\S]|^#([^\s#.,!)]+)$`)
 
 	ModuleAddress = authtypes.NewModuleAddress(ModuleName)
 
 	PostStorePrefix          = []byte("post")
-	PostIndexedIDStorePrefix = []byte("p_index")
-	PostTotalNumberPrefix    = []byte("number_of_posts")
 	PostCommentsStorePrefix  = []byte("comments")
+	PostQueryStorePrefix     = []byte("p_query")
```
Comment (RiccardoM) [inline x/staging/posts/types/keys.go:42]:
```suggestion
	SubspacePostPrefix     = []byte("subspace")
```

File: x/staging/posts/keeper/grpc_query.go (hunk 7 above)
Comment (dadamu) [reply x/staging/posts/keeper/grpc_query.go:39]: @RiccardoM  This is for [`FilteredPaginate`](https://github.com/cosmos/cosmos-sdk/blob/9fd866e3820b3510010ae172b682d71594cd8c14/types/query/filtered_pagination.go#L91), `true` here represents that the result is hit to count the number of element for paginated.

File: x/staging/posts/simulation/utils.go (hunk 10 above)
Comment (dadamu) [reply x/staging/posts/simulation/utils.go:92]: @RiccardoM  Desmos does not filter post with hashtag anymore, so I removed all hashtag related things.

Comment (RiccardoM) [reply x/staging/posts/simulation/utils.go:92]: @dadamu Where they only used inside filtering?

File: x/staging/posts/keeper/grpc_query.go (hunk 7 above)
Comment (RiccardoM) [reply x/staging/posts/keeper/grpc_query.go:39]: @dadamu Oh ok thanks. Great

File: x/staging/posts/simulation/utils.go (hunk 10 above)
Comment (dadamu) [reply x/staging/posts/simulation/utils.go:92]: @RiccardoM Filtering only happens when querying.

Comment (RiccardoM) [reply x/staging/posts/simulation/utils.go:92]: Ok then we can remove them :+1:

File: x/staging/posts/keeper/grpc_query.go (hunk 8 above)
Comment (dadamu) [reply x/staging/posts/keeper/grpc_query.go:20]: @RiccardoM Then, I think this query function should be renaming into `SubspacePosts`, instead.

File: proto/desmos/posts/v1beta1/query.proto (hunk 4 above)
Comment (dadamu) [reply proto/desmos/posts/v1beta1/query.proto]: @RiccardoM Using `subspace_id` instead would be better and explicit.

Comment (RiccardoM) [reply proto/desmos/posts/v1beta1/query.proto]: > @RiccardoM I think using `subspace_id` instead may be better and more explicit. What do you think?

Yeah we can do that

File: x/staging/posts/keeper/grpc_query.go (hunk 8 above)
Comment (RiccardoM) [reply x/staging/posts/keeper/grpc_query.go:20]: @dadamu I prefer we stick with `Posts` instead since there's no need to prepend the `Subspace` prefix as it doesn't add anything to the query. Also, it might confuse developers thinking this will be handled by the `x/subspaces` module

File: docs/developers/queries/staging/posts.md [hunk 13]
```diff
@@ -9,33 +9,19 @@ desmos query posts post [id]
 # desmos query posts post a4469741bb0c0622627810082a5f2e4e54fbbb888f25a4771a5eebc697d30cfc
 ```
 
-# Query the stored posts with filters
-This query allows you to get all the stored posts that match one or more filters. 
+# Query the stored posts inside the subspace
```
Comment (RiccardoM) [inline docs/developers/queries/staging/posts.md:12]:
```suggestion
# Query the posts inside a subspace
```

File: proto/desmos/posts/v1beta1/query.proto [hunk 14]
```diff
@@ -51,49 +51,19 @@ service Query {
...
-    (gogoproto.jsontag) = "parent_id",
-    (gogoproto.moretags) = "yaml:\"id\""
-  ];
-  ;
-  google.protobuf.Timestamp creation_time = 4 [ (gogoproto.stdtime) = true ];
-  string subspace = 5;
-  string creator = 6;
-  repeated string hashtags = 7;
+  // subspace to query the posts for. 
+  // providing an empty value will return an error.
```
Comment (RiccardoM) [inline proto/desmos/posts/v1beta1/query.proto:57]:
```suggestion
  // SubspaceId represents the ID of the subspace to which to query the posts for.
  // Providing an empty or invalid subspace id will return an error.
```

File: x/staging/posts/keeper/grpc_query.go [hunk 15]
```diff
@@ -12,102 +11,47 @@ import (
...
-	var filteredPosts []types.QueryPostResponse
+	var posts []types.Post
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
+	if !subspacestypes.IsValidSubspace(req.SubspaceId) {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "subspace must be a valid sha-256 hash")
```
Comment (RiccardoM) [inline x/staging/posts/keeper/grpc_query.go:25]: I think we can simplify this maybe using `subspacetypes.ErrInvalidSubspaceID` instead: 

```golang
		return nil, sdkerrors.Wrapf(subspacetypes.ErrInvalidSubspaceID, req.SubspaceId)
```

If this variable is not present, we should create it

File: CHANGELOG.md [hunk 16]
```diff
@@ -13,6 +13,7 @@
 - Make `x/profiles` DTag case-insensitive ([#492](https://github.com/desmos-labs/desmos/issues/492))
 - Updated Cosmos SDK to `v0.42.6` ([#509](https://github.com/desmos-labs/desmos/issues/509))
 - Added the possibility to verify a profile with an external application ([#472](https://github.com/desmos-labs/desmos/issues/472)) 
+- Improved the posts query response ([#499](https://github.com/desmos-labs/desmos/issues/499))
```
Comment (RiccardoM) [inline CHANGELOG.md:16]:
```suggestion
- Improved the posts query ([#499](https://github.com/desmos-labs/desmos/issues/499))
```

**Review by RiccardoM — COMMENTED** [review]
Great job, I LOVE seeing so many lines of code deleted 😍 Just a couple of small things and then we're ready to merge this

**Review by RiccardoM — APPROVED** [review]
